crypto/bio/bss_dgram.c

```diff
@@ -294,26 +329,287 @@ static void dgram_reset_rcv_timeout(BIO *b)
 # endif
 }
 
+#if defined(HAVE_IP_PKTINFO)
+/* LINUX has IP_PKTINFO for IPv4 */
+static int dgram_read_unconnected_v4(BIO *b, char *in, int inl,
+                                     int flags,
+                                     BIO_ADDR *dstaddr, BIO_ADDR *peer)
+{
+    int len = 0;
+    unsigned char    chdr[BIO_CMSG_ADDR_SIZE];
+    struct iovec iov;
+    struct msghdr mhdr;
+    struct in_pktinfo *pkt_info = NULL;
+    struct cmsghdr *cmsg;
+    int val;
+
+    /* enable PKTINFO receive */
+    val = 1;
+    if(setsockopt(b->num, IPPROTO_IP, IP_PKTINFO, &val, sizeof(val)) < 0) {
+      return -1;
+    }
+
+    memset(&iov, 0, sizeof(iov));
+    iov.iov_len  = inl;
+    iov.iov_base = (caddr_t) in;
+
+    memset(&mhdr, 0, sizeof(mhdr));
+    mhdr.msg_name = (caddr_t)BIO_ADDR_sockaddr(peer);
+    mhdr.msg_namelen = sizeof(struct sockaddr_in);
+    mhdr.msg_iov = &iov;
+    mhdr.msg_iovlen = 1;
+
+    if(dstaddr != NULL) {
+      memset(chdr, 0, sizeof(chdr));
+      cmsg = (struct cmsghdr *) chdr;
+      mhdr.msg_control = (void *) cmsg;
+      mhdr.msg_controllen = sizeof(chdr);
+    }
+
+    if((len = recvmsg(b->num, &mhdr, 0)) >= 0) {
```

Comment:
Perhaps add `&& dstaddr != NULL` in this expression?

Reply:
I think that the for() loop now has the if dstaddr != NULL that you wanted.
I made it a separate if() statement, because I think that is clearer, but I guess the two lines could be combined.